### PR TITLE
feat(expr): add `regexp_replace`

### LINKED ISSUE #11723: support `regexp_replace`
https://www.postgresql.org/docs/current/functions-matching.html

```
regexp_replace(string, pattern, replacement)

regexp_replace('foobarbaz', 'b..', 'X')
                                   fooXbaz
regexp_replace('foobarbaz', 'b..', 'X', 'g')
                                   fooXX
regexp_replace('foobarbaz', 'b(..)', 'X\1Y', 'g')
                                   fooXarYXazY
```

### PR DESCRIPTION
**Pull request by xzhseh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
resolve #11723

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support `regexp_replace`

</details>

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Please add some e2e tests in `func.slt.part`

File: src/expr/src/expr/expr_regexp.rs
Comment (TennyZhuang) [inline src/expr/src/expr/expr_regexp.rs:475]:
```suggestion
        let global_flag = !flags.is_empty() && flags.starts_with('g');
```

File: src/expr/src/expr/expr_regexp.rs
Comment (TennyZhuang) [inline src/expr/src/expr/expr_regexp.rs:494]: panicked in line 516 if `self.start <= -2`.

Please test the behavior in pg for such case.

I guess the behavior is similar to https://doc.rust-lang.org/std/primitive.usize.html#method.saturating_sub, but you should verify it.

File: src/expr/src/expr/expr_regexp.rs
Comment (TennyZhuang) [inline src/expr/src/expr/expr_regexp.rs:511]: I haven’t understood the purpose of this regular expression yet, but at least the expensive repetitive Regex construction should not be done in match_row. Please move this Regex to RegexpReplaceExpression.

Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:511]: This is used to match the expression like `X\nY` and replace this to `X${n}Y` in order to make use of the `expand` substitution.
But yes, I'll move this to `RegexpReplaceExpression`

**Comment by xzhseh** [comment]
Now we only support `-i`, `-c`, and `-g` flag options, [others](https://www.postgresql.org/docs/current/functions-matching.html#POSIX-EMBEDDED-OPTIONS-TABLE) are not yet supported. The full syntax of `regexp_replace` is of `regexp_replace(source, pattern, replacement [, start [, N ]] [, flags ])`, also note that `start` can not be specified with `flags` if `N` is not specified. (i.e., `regexp_replace('foobarbaz', 'b..', 'X', 1, 'g')` is invalid, and an error will be thrown through parsing stage, which in RW is `try_from`.)

File: src/expr/src/expr/expr_regexp.rs
Comment (wangrunji0408) [inline src/expr/src/expr/expr_regexp.rs:657]: Feels like this branch can be merged with one of the previous branch.

File: e2e_test/batch/basic/func.slt.part
Comment (xzhseh) [reply e2e_test/batch/basic/func.slt.part:475]: The `\&` replacement has not been supported yet, will implement soon.

Comment (xzhseh) [reply e2e_test/batch/basic/func.slt.part:475]: In addition, I'll add the above tests to `e2e_test`, thanks a lot for this!

File: src/expr/src/expr/expr_regexp.rs
Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:490]: Added, thanks for pointing this out!

File: src/expr/src/expr/expr_regexp.rs
Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:519]: Sure, will remove these before merging

File: src/expr/src/expr/expr_regexp.rs
Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:535]: Instead of `replace("\\&", "$0")`, the `$0` should be `${0}` to satisfy the requirement.
Now the `\&` case should be handled.

File: src/expr/src/expr/expr_regexp.rs
Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:538]: I just put the `start` parsing to the very beginning as follows, this should then resolve the unicode index problem.
```
/// Match and replace one row, return the replaced string
fn match_row(&self, text: Option<&str>) -> Option<String> {
  if let Some(text) = text {
      // The start position to begin the search
      let start = if let Some(s) = self.start {
          s - 1
      } else {
          0
      };

      // This is because the source text may contain unicode
      let start = match text.char_indices().nth(start as usize) {
          Some((idx, _)) => idx,
          // With no match
          None => return Some(text.into()),
      };
...
```

File: src/expr/src/expr/expr_regexp.rs
Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:657]: This could be merged, but it will let the overall logic and data flow become ambiguous I think🤔️
Since the difference between this branch and the above one is about the capture group and the use of `expand`.

**Comment by xzhseh** [comment]
New version updated, the left problems may be further reviewed.

**Comment by xzhseh** [comment]
If now the overall picture looks good, I will merge it after removing the debug statement.

File: proto/expr.proto
Comment (xzhseh) [inline proto/expr.proto:105]: Also, is this number identifier `320` in `proto/expr.proto` fits for `regexp_replace`?
I could change this to be compatible if this is not fit.

File: src/expr/src/expr/expr_regexp.rs
Comment (wangrunji0408) [reply src/expr/src/expr/expr_regexp.rs]: I see. I'm also not very familiar with regex syntax. Thanks for correcting me.

File: src/expr/src/expr/expr_regexp.rs
Comment (wangrunji0408) [reply src/expr/src/expr/expr_regexp.rs]: No problem. It's fine as long as the logic is clear.

File: src/expr/src/expr/expr_regexp.rs
Comment (wangrunji0408) [reply src/expr/src/expr/expr_regexp.rs:273]: I think `global_flag` should be put into `RegexpOptions` for consistency. Then we can change `RegexpContext` to:
```rust
pub struct RegexpContext {
    regex: Regex,
    global: bool,
}
```

Btw, should we also enable the [unicode mode](https://docs.rs/regex/latest/regex/struct.RegexBuilder.html#method.unicode) for the regex? 🤔️

File: proto/expr.proto
Comment (wangrunji0408) [reply proto/expr.proto:105]: Maybe `279`? to continue the number within the group.

Comment (xzhseh) [reply proto/expr.proto:105]: Ahh good idea🥰, I directly used the one after `FORMAT` without seeing `278`🤣

File: src/expr/src/expr/expr_regexp.rs
Comment (xzhseh) [reply src/expr/src/expr/expr_regexp.rs:273]: Ahh I see, then it would be more natural to directly use `ctx.global` instead of a separate `global_flag`.
Now the source text with unicode works well (in `e2e_test`) , but I do not know if the unicode mode for `regex` has been enabled or not.

**Comment by xzhseh** [comment]
I've updated the latest version and most of the problems have been resolved.
cc @TennyZhuang  @wangrunji0408, if the overall picture looks good, I'm ready to merge.

**Review by TennyZhuang — APPROVED** [review]
LGTM

File: e2e_test/batch/basic/func.slt.part
Comment (MrCroxx) [inline e2e_test/batch/basic/func.slt.part:547]: 🤣

**Pull request merged**
